### PR TITLE
Pornhub Support

### PR DESCRIPTION
**Pull request by guanw**
Adding support to download videos from pornhub.

### CODE REVIEW

**Comment by guanw** [comment]
Seems like there's a build failure. Can I get some input on how to fix it?

File: extractors/pornhub/pornhub.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
+package pornhub
+
+import (
+	"encoding/json"
```
Comment (iawia002) [inline extractors/pornhub/pornhub.go]: needs a new line after this

File: extractors/pornhub/pornhub.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	"github.com/iawia002/annie/downloader"
+	"github.com/iawia002/annie/request"
+	"github.com/iawia002/annie/utils"
+)
+
+type pornhubData struct {
```
Comment (iawia002) [inline extractors/pornhub/pornhub.go]: go fmt

File: extractors/pornhub/pornhub.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	DefaultQuality bool `json:"defaultQuality"`
+	Format string `json:"format"`
+	Quality string `json:"quality"`
+	VideoUrl string `json:"videoUrl"`
```
Comment (iawia002) [inline extractors/pornhub/pornhub.go]: s/VideoUrl/VideoURL

File: extractors/pornhub/pornhub.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+	if desc != nil {
+		title = desc[1]
+	} else {
+		title = "pornhub video"
+	}
+
+	realURLs := utils.MatchOneOf(html, `"mediaDefinitions":\[(.+?)\],"isVertical"`)
+
+	var pornhubs []pornhubData
+	err = json.Unmarshal([]byte(`[` + realURLs[1] + `]`), &pornhubs)
```
Comment (iawia002) [inline extractors/pornhub/pornhub.go]: Why not directly match the `[]`

File: extractors/pornhub/pornhub.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+	if err != nil {
+		return downloader.EmptyList, err
+	}
+
+	//TODO add support for different quality
+	var realURL string
+	for _, downloadlink := range(pornhubs) {
+		if downloadlink.VideoUrl != "" {
+			realURL = downloadlink.VideoUrl
+			break
```
Comment (iawia002) [inline extractors/pornhub/pornhub.go]: So the first video in the list is the highest quality or?

**Comment by iawia002** [comment]
> Seems like there's a build failure. Can I get some input on how to fix it?

Never mind this, just make sure that the new Pornhub extractor works. I will fix this when I have free time.

File: extractors/pornhub/pornhub.go (hunk 5 above)
Comment (guanw) [reply extractors/pornhub/pornhub.go]: Yes, i think it's sorted like 1080p, 720p and goes on.

File: extractors/pornhub/pornhub.go (hunk 2 above)
Comment (guanw) [reply extractors/pornhub/pornhub.go]: `go fmt` didn't change anything for me somehow.

**Comment by iawia002** [comment]
Hi, @guanw can you do a rebase? There are some conflicts after #444 merged.

**Comment by guanw** [comment]
> Hi, @guanw can you do a rebase? There are some conflicts after #444 merged.

Updated.

File: main.go [hunk 6]
```diff
@@ -150,8 +151,13 @@ func download(videoURL string) {
 		data, err = qq.Extract(videoURL)
 	case "yinyuetai":
 		data, err = yinyuetai.Extract(videoURL)
+<<<<<<< HEAD
 	case "geekbang":
 		data, err = geekbang.Extract(videoURL)
+=======
+	case "pornhub":
+		data, err = pornhub.Extract(videoURL)
+>>>>>>> pornhub support
```
Comment (iawia002) [inline main.go]: remove these

**Comment by guanw** [comment]
My bad... forgot to save it in editor. Updated this time.

**Comment by zhz1237ok** [comment]
oh...................好时代，来临了

**Pull request merged**